diff --git a/src/undetected_chromedriver/patcher.py b/src/undetected_chromedriver/patcher.py
index 5776d19..e3159dd 100644
--- a/src/undetected_chromedriver/patcher.py
+++ b/src/undetected_chromedriver/patcher.py
@@ -334,6 +334,7 @@ class Patcher(object):
         return "".join(cdc).encode()
 
     def is_binary_patched(self, executable_path=None):
+        return True
         executable_path = executable_path or self.executable_path
         try:
             with io.open(executable_path, "rb") as fh:
